extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js

```diff
@@ -0,0 +1,376 @@
+        if (arr[0] !== 'wf') {
+            return {
+                isValid: false,
+                message: `invalid child name: ${arr[0]}`
+            };
+        }
+        if (fd.workflowId === undefined ||
+            fd.workflowVersion === undefined ||
+            fd.type === undefined ||
+            fd.subType === undefined ||
+            fd.nextNodes === undefined) {
+            return {
+                isValid: false,
+                message: 'missing some fields'
+            };
+        }
+        if (fd.subType === this.wed.SUB_TYPE_BASIC) {
+            if (fd.bucket === undefined) {
+                return {
+                    isValid: false,
+                    message: 'basic: missing bucket'
+                };
+            }
+        }
+        if (fd.workflowId !== arr[1]) {
+            return {
+                isValid: false,
+                message: `non matching workflow id: ${arr[1]}`
+            };
+        }
+        // eslint-disable-next-line
+        if (fd.workflowVersion != arr[2]) {
+            return {
+                isValid: false,
+                message: `non matching workflow version: ${arr[2]}`
+            };
+        }
+        this.filterDescriptors[name] = fd;
+        return {
+            isValid: true
+        };
+    }
+
+    // for tests
+    _getFilterDescriptorsLength() {
+        return Object.keys(this.filterDescriptors).length;
+    }
+
+    // for tests
+    _deleteAllFilterDescriptors() {
+        this.filterDescriptors = [];
+    }
+
+    // check if zkClient exists and is connected before
+    // (for timer)
+    __updateFilterDescriptors() {
+        if (!this.zkClient) {
+            this.log.debug('WEDSQP: zk not initialized');
+            return;
+        }
+        if (this.zkClient.getState !== NodeZookeeperClient.State.CONNECTED) {
+            this.log.debug('WEDSQP: zk not connected');
+            return;
+        }
+        this._updateFilterDescriptors();
+    }
+
+    // update the filter descriptor table acc/to new information
+    // cb is optional
+    _updateFilterDescriptors(cb) {
+        const { zookeeperPath } = this.extConfig;
+        this.log.debug(
+            `WEDSQP: _updateFilterDescr ${this.filterDescriptors.length}`);
+        this.zkClient.getChildren(
+            zookeeperPath,
+            (err, children) => {
+                if (err) {
+                    this.log.error(
+                        'zookeeper could not get children', {
+                            method:
+                            'WEDSQP._updateFilterDescriptors',
+                            error: err,
+                        });
+                    if (cb) {
+                        return cb(err);
+                    }
+                    return undefined;
+                }
+                this.log.debug(
+                    `WEDSQP: reading children data ${children.length}`);
+                async.times(children.length, (i, next) => {
```

Comment:
a more idiosynchratic function to do the same: `async.map(children, (child, next) => ..., (err, results) => {...})`

You might want to use `mapSeries` or `mapLimit` to control parallelism better.